docs/middleware/BLSPublicKeyCompendium.md

```diff
@@ -0,0 +1,23 @@
+# BLSPublicKeyCompendium
+
+This contract is shared by all AVSs and serves a single place for operators to connect their execution layer address to a bn254 public key. When operators opt into AVSs, the AVS contracts can read from the BLSPublicKeyCompendium in order to get the BLS public key of the operator. This contract also prevents against [rogue key attacks](https://xn--2-umb.com/22/bls-signatures/#rogue-key-attack).
+
+## Flows
+
+There is only one flow for this contract, which is a call from an operator to register their execution layer address with thier bn254 public key.
+
+### Cryptographic Verification
+
+Operators provide the contract with 
+- A BLS signature $\sigma \in \mathbb{G}_1$ of $M = (msg.sender, chain.id, \text{"EigenLayer\_BN254\_Pubkey\_Registration"})$
+- Their public keys $pk_1 \in \mathbb{G}_1$ and $pk_2 \in \mathbb{G}_2$
+
+The contract then 
+- Calculates $\gamma = keccak256(\sigma, pk_1, pk_2, M)$
+- Verifies $e(\sigma + \gamma pk_1, [1]_2) = e(H(m) + \gamma[1]_1, pk_2)$
+
+This verifies that the operator owns the secret key corresponding to the public keys and that the $pk_1$ and $pk_2$ have the same discrete logarithm according to their respective curve's generators.
+
+### Integrations
+
+The contract stores a map from the execution layer address to the hash of the $\mathbb{G}_1$ public key and the other way around. This means that the $\mathbb{G}_1$ public key needs to sent in calldata onchain, hashed, and verified when AVSs want to do operations with individual operator public keys (remember to link here).
```

Comment:
I still think the one-time-per-operator cost of also storing the entire pubkey instead of just the hash would be worth it, to support applications where providing the key in calldata is infeasible or could run into race conditions.

there is also a typo here: `needs to sent` should be `needs to be sent` (missing 'be')

Reply:
Fixed cf883904

Can you give an example of where just emitting the event would not be ok?

Reply:
> Can you give an example of where just emitting the event would not be ok?

somewhat theoretical, but if I have a list of all operators in my AVS and want to rotate them out (e.g. churn out the one with the lowest stake), then either I need a bunch of off-chain infra to help keep track of keys, provide them to newly registering operators, etc. _or_ I could just have the full keys on-chain to lookup when needed

Reply:
The hope is that it is not a bunch of offchain infra

Reply:
the other advantage here is you can more elegantly deal with some race conditions.
e.g. if we can find the correct operator on-chain (at runtime) but need the preimage, then a transaction can revert if a another transaction changes who the "correct operator" is (tx reverts since the provided preimage will now be wrong), but if we can just look up the correct key then we can resolve this race condition

Reply:
FWIW, @cryptogakusei also raised a question about if this storage would be added. Perhaps he can provide some context on his thoughts here